rfcs/rdm-0045-versioning-support.md

```diff
@@ -0,0 +1,530 @@
+
+Having ownership/permissions managed at a single version level will lead to strange issues, like an owner of a record version not being able to manage prior versions (this issue has e.g. been seen in the GitHub integration on Zenodo). Also, it becomes very complex to assign permissions to other users. For instance, an owner that want to remove access for a user has to go to all versions (potentially several hundreds).
+
+The community integration works in a similar way to owners, in that all versions or none belongs to a community.
+
+**Ownership and permissions takes effect immediately**
+
+A key design decision is if ownership and permissions is managed differently on records vs drafts. Drafts are being edited, and then published to make the changes available.
+
+Thus, the question is when is ownership and permissions taking effect. For instance, an owner gives view permission on an already published record to a user via the deposit form. What happens?
+
+- Immediate effect:  If the changes take effect immediately, the user can now view the record.
+- On publish:  If the changes take effect on publish, then the user cannot view the record until the new record is published.
+
+Having permissions take effect immediately opens other use cases of directly sharing a record from the landing page instead of going through the deposit form. Also the model seems simpler to understand for end-users.
+
+### Exclusion: Version relationships across repositories
+
+You could imagine modelling version relationships across repositories. For instance a dataset is first uploaded to repository X, then later a new version is published in repository Y.
+
+We will not complicate the versioning support with this style of relationships. Relationships across repositories should be modelled in metadata via the related identifiers field. In addition, another workaround is to transfer the full dataset from repository X into repository Y.
+
+## Backend Design
+
+The backend consist of the data access, service and presentation layers.
+
+### Data access layer
+
+#### Version relationships
+
+We propose that the information previously held by Invenio-PIDRelations is moved directly to the record database table. The current attributes of the ``RDMRecordMetadata`` model includes:
+
+- ``id `` - UUID of the record
+- ``json`` - Record JSON document
+- ``version_id`` - Version counter used for optimistic concurrency control.
+- ``created `` - Creation timestamp.
+- ``updated`` - Modification timestamp.
+- ``bucket_id`` - Foreign key (UUID) to a Files-REST bucket.
+
+We propose to add the following two attributes:
+
+- ``parent_id`` - Foreign key (UUID) to a Concept Record (a new entity).
+- ``parent_order`` - Integer (4 bytes) defining the order of siblings.
+- ``parent_latest`` - Boolean to indicate if the record is the latest version.
```

Comment:
I know commits have already been merged, but here are my takes on the names:

``parent_id`` is both the id of the parent *and* namespacing for the "versioning" features. This makes ``parent_order`` and ``parent_latest`` confusing because ``parent`` there is only for namespacing. How about: 

```suggestion
- ``versioning_parent_id`` / ``versioning_concept_id`` - Foreign key (UUID) to a Concept Record (a new entity).
- ``versioning_order`` - Integer (4 bytes) defining the rank of the record among sibling versions.
- ``versioning_latest`` - Boolean to indicate if the record is the latest version.
```